ostd/src/sync/wait.rs

```diff
@@ -242,6 +242,32 @@ impl Drop for Waiter {
 }
 
 impl Waker {
+    /// Creates a `Waker` for waking up the target task.
+    /// It needs to create a corresponding [`Waiter`] on the target task thus
+    /// the waker can work in coordination with the waiter.
+    pub fn new(task: Arc<Task>) -> Arc<Self> {
+        Arc::new(Self {
+            has_woken: AtomicBool::new(true),
+            task,
+        })
+    }
+
+    /// Tries to create a [`Waiter`] associated with this waker.
+    ///
+    /// If this method is called within the task corresponding to the waker,
+    /// it will successfully create the corresponding waiter and return it.
+    /// Otherwise, it will return `None`.
+    pub fn try_create_waiter(self: &Arc<Self>) -> Option<Waiter> {
```

Comment:
Creating multiple waiters for the same waker will mess up the state, and this API cannot prevent that at all.

I still think it is better to keep `SpinLock<Option<Waker>>` in `Thread` and just install the waker before checking the pending signals and entering sleep. This has some overhead due to the spin lock, but I think it's more correct.

Any thoughts on this?

Reply:
I am still trying to find a relatively reasonable method to avoid locking the waker in a thread, as even minor performance overheads are currently a concern. 

I am not entirely clear on what constitutes a "mess up" of the state. I feel that some changes in the state might introduce additional overhead to the original execution, but they are unlikely to affect the final correctness. This is also occurs in the current APIs. For example, a thread creates a pair of waiter and waker, and the waker is passed to another thread. If the other thread calls the waker's `wake_up` before the waiter performs a wait, the original state will be set to true. Subsequent waits will result in an extra loop, but this does not affect the final correctness. I am unsure if there is currently an invariant definition for `has_woken`?

Reply:
> I am still trying to find a relatively reasonable method to avoid locking the waker in a thread, as even minor performance overheads are currently a concern.

Are you sure that the overhead of a simple spinlock has a significant (or at least observable) impact? Do you have any data to support your argument? For example, if it's only 1% in your benchmark, it shouldn't matter.

> I am unsure if there is currently an invariant definition for has_woken?

The invariant is that it _never_ loses a wakeup event. But this invariant is clearly broken by the newly added `try_create_waiter` method.
 - `try_create_waiter` sets `has_woken` to false, so if we already have another waiter for the same waker, and the waiter has been woken, the wakeup event will be lost.
 - Even without the above fact, after the second waiter finishes waiting, it will also set `has_woken` to false, so, again, the wakeup event for the first waiter will be lost.

We cannot really allow two waiters to be created for the same waker, otherwise if someone wakes the waker, we will never know which waiter should be woken.

> This is also occurs in the current APIs. For example, a thread creates a pair of waiter and waker, and the waker is passed to another thread. If the other thread calls the waker's wake_up before the waiter performs a wait, the original state will be set to true. Subsequent waits will result in an extra loop, but this does not affect the final correctness.

That's fine. No wakeup event is lost.

> I am not entirely clear on what constitutes a "mess up" of the state. I feel that some changes in the state might introduce additional overhead to the original execution, but they are unlikely to affect the final correctness.

While calling `wait_until`, we can call `wait_until` again in the `cond` closure.
 - If the nested `wait_until` accidentally reuses the waiter and waker of the outer `wait_until` (in other words, if they're both thread-directed waiter and waker), some wakeup events for the outer `wait_until` may be lost, and there is a probability that it will never be woken up.

So the final correctness is also broken.

Reply:
> While calling wait_until, we can call wait_until again in the cond closure.

Considering the nested situation, if we store `SpinLock<Option<Arc<Waker>>>` and set the waker once we use a signalled waiter (Pauser in this PR) to do wait, should we only allow one signalled waiter exists to avoid the nested situation? Otherwise the inner waiter will replace the waker set by outer waiter.

Reply:
> set the waker once we use a signalled waiter (Pauser in this PR)

I don't think we should do this.

Installing the waker on the thread should be done in `cancel_cond` (the same place where we should check for pending signals, see the comment above), so I think the signal waker will never be nested.